File: runtimes/browser-automation/src/browser/artifacts.ts
import { mkdir, realpath } from "node:fs/promises";
import path from "node:path";

export class BrowserArtifactRoot {
  readonly root: string;
  readonly uploadRoots: string[];

  constructor(root: string, options?: { uploadRoots?: string[] }) {
    this.root = path.resolve(root);
    this.uploadRoots = (options?.uploadRoots ?? []).map((entry) => path.resolve(entry));
  }

  outputPath(kind: "screenshots" | "downloads" | "uploads" | "traces" | "pdf", fileName: string) {
    if (fileName !== path.basename(fileName)) {
      throw new Error("outside artifact root");
    }
    const resolved = path.resolve(this.root, kind, fileName);
    assertInsideRoot(resolved, path.join(this.root, kind), "outside artifact root");
    return resolved;
  }

  async ensureKind(kind: "screenshots" | "downloads" | "uploads" | "traces" | "pdf") {
    await mkdir(path.join(this.root, kind), { recursive: true });
  }

  async inputUploadPath(filePath: string): Promise<string> {
    const resolved = path.resolve(filePath);
    if (!this.uploadRoots.some((root) => isInsideRoot(resolved, root))) {
      throw new Error("outside upload roots");
    }
    const real = await realpath(resolved);
    if (!this.uploadRoots.some((root) => isInsideRoot(real, root) || isInsideRoot(resolved, root))) {
      throw new Error("outside upload roots");
    }
    return real;
  }
}

function assertInsideRoot(candidate: string, root: string, message: string): void {
  if (!isInsideRoot(candidate, root)) {
    throw new Error(message);
  }
}

function isInsideRoot(candidate: string, root: string): boolean {
  const relative = path.relative(root, candidate);
  return relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative));
}

File: crates/browser-automation/src/artifacts.rs
use crate::{BrowserAutomationError, BrowserResult};
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BrowserArtifactRoot {
    root: PathBuf,
    upload_roots: Vec<PathBuf>,
}

impl BrowserArtifactRoot {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self {
            root: root.into(),
            upload_roots: Vec::new(),
        }
    }

    pub fn with_upload_roots(mut self, upload_roots: Vec<PathBuf>) -> Self {
        self.upload_roots = upload_roots;
        self
    }

    pub fn output_path(&self, kind: &str, file_name: &str) -> BrowserResult<PathBuf> {
        if Path::new(file_name)
            .file_name()
            .and_then(|value| value.to_str())
            != Some(file_name)
        {
            return Err(BrowserAutomationError::ArtifactPath(
                "outside artifact root".to_string(),
            ));
        }
        let path = self.root.join(kind).join(file_name);
        if !path.starts_with(self.root.join(kind)) {
            return Err(BrowserAutomationError::ArtifactPath(
                "outside artifact root".to_string(),
            ));
        }
        Ok(path)
    }

    pub fn validate_upload_path(&self, path: &Path) -> BrowserResult<()> {
        if self.upload_roots.iter().any(|root| path.starts_with(root)) {
            Ok(())
        } else {
            Err(BrowserAutomationError::ArtifactPath(
                "outside upload roots".to_string(),
            ))
        }
    }
}
